consensus/core/src/leader_scoring.rs

```diff
@@ -1,12 +1,120 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use std::{cmp::Ordering, fmt::Debug, sync::Arc};
+use std::{
+    cmp::Ordering,
+    collections::{BTreeMap, HashMap, HashSet},
+    fmt::Debug,
+    ops::Bound::{Excluded, Included},
+    sync::Arc,
+};
 
 use consensus_config::AuthorityIndex;
 use serde::{Deserialize, Serialize};
+use itertools::Itertools;
 
-use crate::{commit::CommitRange, context::Context};
+use crate::{
+    block::{BlockAPI, BlockDigest, BlockRef, Slot, VerifiedBlock},
+    commit::CommitRange,
+    context::Context,
+    leader_scoring_strategy::ScoringStrategy,
+    stake_aggregator::{QuorumThreshold, StakeAggregator},
+    universal_committer::UniversalCommitter,
+    CommittedSubDag, Round,
+};
+
+#[allow(unused)]
+pub(crate) struct ReputationScoreCalculator<'a> {
+    // The range of commits that these scores are calculated from.
+    pub(crate) commit_range: CommitRange,
+    // The scores per authority. Vec index is the `AuthorityIndex`.
+    pub(crate) scores_per_authority: Vec<u64>,
+
+    // As leaders are sequenced the subdags are collected and cached in `DagState`.
+    // Then when there are enough commits to trigger a `LeaderSchedule` change,
+    // the subdags are then combined into one `UnscoredSubdag` so that we can
+    // calculate the scores for the leaders in this subdag.
+    unscored_subdag: UnscoredSubdag,
+    // There are multiple scoring strategies that can be used to calculate the scores
+    // and the `ReputationScoreCalculator` is responsible for applying the strategy.
+    // For now this is dynamic while we are experimenting but eventually we can
+    // replace this with the final strategy that works best.
+    scoring_strategy: Box<dyn ScoringStrategy>,
+    // We use the `UniversalCommitter` to elect the leaders from the `UnscoredSubdag`
+    // that need to be scored.
+    committer: &'a UniversalCommitter,
+}
+
+#[allow(unused)]
+impl<'a> ReputationScoreCalculator<'a> {
+    pub(crate) fn new(
+        context: Arc<Context>,
+        committer: &'a UniversalCommitter,
+        unscored_subdags: &Vec<CommittedSubDag>,
+        scoring_strategy: Box<dyn ScoringStrategy>,
+    ) -> Self {
+        let num_authorities = context.committee.size();
+        let scores_per_authority = vec![0_u64; num_authorities];
+
+        assert!(
+            !unscored_subdags.is_empty(),
+            "Attempted to calculate scores with no unscored subdags"
+        );
+
+        let unscored_subdag = UnscoredSubdag::new(context.clone(), unscored_subdags);
+
+        let commit_indexes = unscored_subdags
+            .iter()
+            .map(|subdag| subdag.commit_index)
+            .collect::<Vec<_>>();
+        let (min_commit_index, max_commit_index) =
+            commit_indexes.iter().minmax().into_option().unwrap();
```

Comment:
Changed it so that when an UnscoredSubdag is created the list of subdags that are passed in will be checked to ensure that we have a contiguous (and therefore sorted) list of commit indices. Then we will also have the blocks collected in order as well so we can get the commit range and leader range easily from the UnscoredSubdag.

> The result from minmax() is inclusive for both start and end. So CommitRange needs to end on *max_commit_index+1 exclusive.

I am unsure what you meant by this, so I did not make any changes here. The commit range is used for debugging and we consider the bounds of the range inclusive. Even when we store this information in the next PR we store it along side the last commit reference.

https://github.com/MystenLabs/sui/blob/5694132b59b366455eeb5db5967a703c55e2be9f/consensus/core/src/dag_state.rs#L596-L610

Reply:
Even if it is only used for debugging, it is strange that CommitRange is wrapping the Range type that has inclusive start and exclusive end, while we are using values with inclusive start and end. For example, if the first round among blocks is 5 and last round among blocks is 10, the correct CommitRange should be 5..11 or 5..=10, instead of 5..10.

Reply:
Aaah I see what you mean now, modified the related code to reflect this inclusive/exclusive range.